### PR TITLE
GitHub tarball support

### PR DESCRIPTION
**Pull request by rodjek**
Fleshes out the GitHub tarball support.

All methods are TomDoc'd and all the changes have 100% test coverage.

### CODE REVIEW

**Comment by wfarr** [comment]
I still need to pour over the code a little bit, but I'd love to just use `github` instead of `github_tarball` I think (though we can support `github_tarball` for legacy reasons).

I also would like to see the `github` helper method we use in Boxen formalized in the API to be a short-hand for `mod "whatever", "version", :github => "blah"` like so:

``` ruby
# stores in $path/puppetlabs-stdlib
github "puppetlabs/puppetlabs-stdlib", "~> 4.0"

github "stdlib", "~> 4.0", :repo => "puppetlabs/puppetlabs-stdlib"
```

What do you think about that?

**Comment by rodjek** [comment]
As per the spec in the integration test, currently it only supports the following syntax

``` ruby
mod 'stdlib', '~> 2', :github => 'puppetlabs/stdlib'
```

That said, any additional Puppetfile DSL changes are trivial to make.

File: lib/henson.rb [hunk 1]
```diff
@@ -38,4 +38,8 @@ class RequirementNotSatisfied < Error; exit_code(20); end
   class GitNotInstalled         < Error; exit_code(30); end
   class GitRefNotFound          < Error; exit_code(32); end
   class GitInvalidRef           < Error; exit_code(34); end
+
+  class GitHubTarballNotFound   < Error; exit_code(40); end
```
Comment (wfarr) [inline lib/henson.rb:42]: Maybe `GitHubInvalidRelease` or similar here?

File: lib/henson.rb [hunk 2]
```diff
@@ -38,4 +38,8 @@ class RequirementNotSatisfied < Error; exit_code(20); end
...
+  class GitHubAPIError          < Error; exit_code(42); end
```
Comment (wfarr) [inline lib/henson.rb:43]: We may want to differentiate between `GitHubInvalidAPIRequest` and `GitHubInvalidAPIResponse`?

File: lib/henson/dsl/puppetfile.rb [hunk 3]
```diff
@@ -30,8 +30,12 @@ def mod(name, *args)
         options = args.last.is_a?(Hash) ? args.pop : {}
         version = args.empty? ? ">= 0" : args.first
 
-        unless options.any? && forge.nil?
-          options.merge!(:forge => forge)
+        unless options.any?
+          if forge.nil?
+            # TODO - Implicit forge URL or throw error?
```
Comment (wfarr) [inline lib/henson/dsl/puppetfile.rb:35]: I'm +:100: saying a forge declaration of some kind is required.

File: lib/henson/dsl/puppetfile.rb [hunk 4]
```diff
@@ -47,6 +51,28 @@ def forge(url = nil)
...
+      def github(name, *args)
+        options = args.last.is_a?(Hash) ? args.pop : {}
+        version = args.empty? ? ">= 0" : args.first
+
+        unless name =~ /\A.+\/.+\Z/
+          raise ModulefileError, "'#{name}' is not a GitHub repository"
+        end
+
+        if options[:repo]
+          options[:github] = options[:repo]
```
Comment (wfarr) [inline lib/henson/dsl/puppetfile.rb:64]: This can be condensed:

``` ruby
options[:github] = options.delete(:repo) || name
```

File: lib/henson/dsl/puppetfile.rb [hunk 5]
```diff
@@ -47,6 +51,28 @@ def forge(url = nil)
...
+          options.delete(:repo)
+        else
+          options[:github] = name
+        end
+
+        module_name = name.split('/').last.gsub(/\Apuppet(labs)?-/, '')
```
Comment (wfarr) [inline lib/henson/dsl/puppetfile.rb:70]: I like the intelligent module name selection behavior here.

File: lib/henson/dsl/puppetfile.rb [hunk 6]
```diff
@@ -47,6 +51,28 @@ def forge(url = nil)
...
+
+        PuppetModule.new(module_name, version, options).tap do |puppet_module|
+          @modules << puppet_module
```
Comment (wfarr) [inline lib/henson/dsl/puppetfile.rb:73]: If we've only got the one module here, and no in-line modifications, `#tap` is just going to incur a performance hit (blocks vs. inlining, etc) without any readability benefits _in this particular case_, IMO.

Would rather see:

``` ruby
@modules << PuppetModule.new module_name, version, options
```

File: lib/henson.rb (hunk 1 above)
Comment (rodjek) [inline lib/henson.rb:42]: If I request a module that doesn't exist, NotFound seems a better fit than InvalidRelease.

File: lib/henson/settings.rb [hunk 7]
```diff
@@ -5,7 +5,8 @@ def initialize
         :quiet      => false,
         :verbose    => false,
         :puppetfile => "#{Dir.pwd}/Puppetfile",
-        :path       => "#{Dir.pwd}/shared"
+        :path       => "#{Dir.pwd}/shared",
+        :cache_path => "#{Dir.pwd}/.henson/cache",
```
Comment (wfarr) [inline lib/henson/settings.rb:9]: We can revisit this with another PR, but I want to take a Bundler-like approach to this:
- If `--path` is not given or specified otherwise in the config, use `~/.puppet/henson/cache`
- Otherwise, keep the cache in `--path` if cache is enabled in config or via cli opt

File: lib/henson/source/generic.rb [hunk 8]
```diff
@@ -10,7 +10,7 @@ def versions
       end
 
       def resolve_version_from_requirement(requirement)
-        satisfiable_versions_for_requirement(requirement).sort.first
+        satisfiable_versions_for_requirement(requirement).sort.reverse.first
```
Comment (wfarr) [inline lib/henson/source/generic.rb:13]: Unnecessary array duplication here.

We could make use of `Enumerable#max` here maybe.

File: lib/henson/dsl/puppetfile.rb (hunk 3 above)
Comment (rodjek) [inline lib/henson/dsl/puppetfile.rb:35]: Yeah, I was leaning towards that myself but I thought I'd wait for a consensus.

File: lib/henson/source/github_tarball.rb [hunk 9]
```diff
@@ -1,9 +1,223 @@
...
+
+      private
+
+      # Internal: Make a call to the GitHub API.
+      #
+      # path - The URI String under https://api.github.com to request (e.g.
+      # /repos/puppetlabs/puppetlabs-stdlib/tags).
+      #
+      # Returns the parsed JSON object (probably a Hash).
+      def api_call(path)
```
Comment (wfarr) [inline lib/henson/source/github_tarball.rb:90]: I absolutely want the API client pulled out into a separate class. This class has too many responsibilities atm.

File: spec/henson/dsl/puppetfile_spec.rb [hunk 10]
```diff
@@ -94,4 +94,43 @@
       expect(instance.instance_variable_get("@forge")).to eq("foobar")
     end
   end
+
+  context "#github" do
+    let(:mod) { instance.github('puppetlabs/puppetlabs-stdlib', '~> 1') }
+
+    before do
+      FakeWeb.register_uri(
```
Comment (wfarr) [inline spec/henson/dsl/puppetfile_spec.rb:102]: So one of the benefits of refactoring the API client into another class is FakeWeb can be eliminated entirely. :grin:

File: lib/henson/settings.rb (hunk 7 above)
Comment (rodjek) [inline lib/henson/settings.rb:9]: Definitely something that deserves its own PR.

File: lib/henson.rb (hunk 1 above)
Comment (wfarr) [inline lib/henson.rb:42]: I agree the message presented to the user should mention that the cause is that the release could not be found _because_ it does not exist.

I think that `GitHubInvalidRelease` should be considered more internal terminology.

File: lib/henson/dsl/puppetfile.rb (hunk 3 above)
Comment (wfarr) [inline lib/henson/dsl/puppetfile.rb:35]: :sparkles:

File: lib/henson/source/generic.rb (hunk 8 above)
Comment (rodjek) [inline lib/henson/source/generic.rb:13]: `#max` would work but we should probably consider returning `Gem::Version` objects from our `Source#versions` method rather than strings.

File: lib/henson/settings.rb (hunk 7 above)
Comment (wfarr) [inline lib/henson/settings.rb:9]: :+1:

File: lib/henson/source/generic.rb (hunk 8 above)
Comment (wfarr) [inline lib/henson/source/generic.rb:13]: `Enumerable#max` takes a block. We can absolutely have the return value be a `Gem::Version`.

The other thing I'm rapidly considering _over_ RubyGems here is relying on [solve](https://github.com/RiotGames/solve) instead for both constraints on individual versions and constraints for dependency resolution.

File: lib/henson/source/github_tarball.rb [hunk 11]
```diff
@@ -1,9 +1,223 @@
...
+      #
+      # tarball - The String path on disk to the tarball.
+      # dest    - The String path on disk to the directory that the tarball
+      #           will be extracted into.
+      #
+      # Returns nothing.
+      def extract_tarball(tarball, dest)
+        Henson.ui.debug "Extracting #{tarball} to #{dest}"
+
+        Gem::Package::TarReader.new(Zlib::GzipReader.open(tarball)).each do |entry|
```
Comment (wfarr) [inline lib/henson/source/github_tarball.rb:181]: IIRC, @mitchellh was doing stuff with Zlib on Vagrant to make it suck a bit less to use. I'd love to hear some more about that from him if we can.

File: lib/henson.rb (hunk 1 above)
Comment (rodjek) [inline lib/henson.rb:42]: I would agree if that error was being raised by anything that actually checked if a release was valid. Even internally, InvalidRelease suggests that the module exists but the requested version/release doesn't. That's not what this error represents. We can always add an InvalidRelease error if needed later on.

Comment (wfarr) [inline lib/henson.rb:42]: Perhaps two different errors? `GitHubRepositoryDoesNotExist` and `GitHubReleaseDoesNotExist`?

File: lib/henson/source/github_tarball.rb (hunk 11 above)
Comment (rodjek) [inline lib/henson/source/github_tarball.rb:181]: It's not really that sucky in its current form, it Just Works (tm) for what we need it to do.

File: lib/henson/source/github_tarball.rb (hunk 9 above)
Comment (rodjek) [inline lib/henson/source/github_tarball.rb:90]: I don't see the usefulness of that, we can revisit it later if it becomes necessary.

File: spec/henson/dsl/puppetfile_spec.rb (hunk 10 above)
Comment (rodjek) [inline spec/henson/dsl/puppetfile_spec.rb:102]: I don't see how that'd be possible without removing the unit tests, which isn't going to happen.

File: lib/henson/dsl/puppetfile.rb (hunk 4 above)
Comment (rodjek) [inline lib/henson/dsl/puppetfile.rb:64]: good call

File: lib/henson/dsl/puppetfile.rb (hunk 5 above)
Comment (rodjek) [inline lib/henson/dsl/puppetfile.rb:70]: Yep, 99.99% of modules are named puppet-foo or puppetlabs-foo, so I thought it was fair to enforce that naming scheme in order to use this helper.

File: lib/henson/source/github_tarball.rb (hunk 11 above)
Comment (mitchellh) [inline lib/henson/source/github_tarball.rb:181]: @wfarr Ah, what I'm doing probably won't be of interest to you here. I got tired of the crappy performance and error handling of the pure Ruby tar/zlib stuff, so now I package `bsdtar` with Vagrant and shell out to it.

**Comment by wfarr** [comment]
It's 4AM Sydney time. Merging this and tackling all the niggles as secondary PRs.
